### PR TITLE
Validate booking slots

### PR DESCRIPTION
**Pull request by kpuputti**
This PR checks that the booking dates (between start date inclusive, end date exclusive) are available in the Harmony API before moving into payment.

## TODO

- [X] Date query param encoder to HarmonyClient
- [x] Query timeslots from Harmony
- [x] Validate all days/nights are available in Harmony response
- [x] Only validate timeslots with feature flag and availability turned on
- [x] Git cleanup

### CODE REVIEW

File: app/controllers/preauthorize_transactions_controller.rb [hunk 1]
```diff
@@ -146,6 +159,52 @@ def validate_booking(tx_params:, quantity_selector:)
...
 
+    def all_days_available(timeslots, start_on, end_on)
+      # Take all the days except the exclusive last day
+      requested_days = (start_on..end_on).to_a[0..-2]
+
+      available_days = timeslots
+                         .select { |s| s[:unitType] == :day }
+                         .map { |s| s[:start].to_date }
+
+      requested_days.all? { |d|
```
Comment (rap1ds) [inline app/controllers/preauthorize_transactions_controller.rb]: This is perfectly fine, no need to change, but If you want to make this without iterating, you could turn these arrays to Sets and use set operations.

File: app/utils/service_client/middleware/param_encoder.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    class ParamEncoder < MiddlewareBase
+      def enter(ctx)
+        params = ctx[:req][:params]
+        ctx[:req][:params] = encode(params)
+        ctx
+      end
+
+      private
+
+      def encode(params)
```
Comment (kpuputti) [inline app/utils/service_client/middleware/param_encoder.rb]: Thanks for a great suggestion! The `is_a?` felt awkward but I didn't know any better. Your suggestion sounds good, I'll change to that.

File: config/locales/en.yml [hunk 3]
```diff
@@ -1599,6 +1599,8 @@ en:
       night: night
       nights: nights
       invalid_parameters: "Invalid values for new transaction"
+      dates_not_available: "Selected dates not available"
```
Comment (rap1ds) [inline config/locales/en.yml]: Is this missing _are_ ?

File: app/services/result.rb [hunk 4]
```diff
@@ -96,6 +114,14 @@ def and_then(&block)
       self
     end
 
+    # Error a -> Result b
+    def rescue(&block)
+      result = block.call(self)
```
Comment (rap1ds) [inline app/services/result.rb]: The Result object is now passed to the handler block is as, which is different what `and_then` does. `and_then` passes unwrapped `data` to the handler. So I think `rescue` should also pass unwrapped `data` and the `error_msg` the same way as `on_error` does: https://github.com/sharetribe/sharetribe/pull/2559/files#diff-3806624eafa2c3ce0e7273410ae8e337R131

Does it make sense?

**Pull request merged**